chapter_optimization/minibatch-sgd.md

```diff
@@ -272,6 +348,40 @@ def train_ch11(trainer_fn, states, hyperparams, data_iter,
     return timer.cumsum(), animator.Y[0]
 ```
 
+```{.python .input}
+#@tab tensorflow
+#@save
+def train_ch11(trainer_fn, states, hyperparams, data_iter,
+               feature_dim, num_epochs=2):
+  # Initialization
+  w = tf.Variable(tf.random.normal(shape=(feature_dim, 1),
+                                   mean=0, stddev=0.01),trainable=True)
+  b = tf.Variable(tf.zeros(1), trainable=True)
+  
+  # Train
+  net, loss = lambda X: d2l.linreg(X, w, b), d2l.squared_loss
+  animator = d2l.Animator(xlabel='epoch', ylabel='loss',
+                            xlim=[0, num_epochs], ylim=[0.22, 0.35])
+  n, timer = 0, d2l.Timer()
+
+  for _ in range(num_epochs):
+    for X, y in data_iter:
+      with tf.GradientTape() as g:
+        l = tf.math.reduce_mean(loss(net(X), y))
+      
+      dw, db = g.gradient(l, [w, b])
+      trainer_fn([w, b], [dw, db], states, hyperparams)
+      n += X.shape[0]
+      if n % 200 == 0:
+          timer.stop()
+          p = n/X.shape[0]/tf.data.experimental.cardinality(data_iter).numpy()
+          q = (d2l.evaluate_loss(net, data_iter, loss),)
+          animator.add(p, q)
+          timer.start()
+  print(f'loss: {animator.Y[0][-1]:.3f}, {timer.avg():.3f} sec/epoch')
+  return timer.cumsum(), animator.Y[0]
```

Comment:
Please use appropriate tab spacing (4) for the function.

Reply:
please check now.